execution/block-partitioner/src/v2/mod.rs

```diff
@@ -0,0 +1,719 @@
+                        .chain(wsets[txn_idx].read().unwrap().iter())
+                    {
+                        let key_idx = *key_idx_ref;
+                        let tracker = trackers.get(&key_idx).unwrap();
+                        tracker.write().unwrap().mark_txn_ordered(
+                            txn_idx,
+                            last_round_id,
+                            num_executor_shards - 1,
+                        );
+                    }
+                });
+            });
+
+            remaining_txns = vec![vec![]; num_executor_shards];
+            remaining_txns[num_executor_shards - 1] = last_round_txns;
+            txn_idx_matrix.push(remaining_txns);
+            let duration = timer.stop_and_record();
+            info!("last_round={duration}");
+        }
+
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["new_tid_table"])
+            .start_timer();
+        let num_rounds = txn_idx_matrix.len();
+        let mut start_index_matrix: Vec<Vec<TxnIndex>> =
+            vec![vec![0; num_executor_shards]; num_rounds];
+        let mut global_counter: TxnIndex = 0;
+        for (round_id, row) in txn_idx_matrix.iter().enumerate() {
+            for (shard_id, txns) in row.iter().enumerate() {
+                start_index_matrix[round_id][shard_id] = global_counter;
+                global_counter += txns.len();
+            }
+        }
+
+        self.thread_pool.install(|| {
+            (0..num_rounds).into_par_iter().for_each(|round_id| {
+                (0..num_executor_shards)
+                    .into_par_iter()
+                    .for_each(|shard_id| {
+                        let sub_block_size = txn_idx_matrix[round_id][shard_id].len();
+                        (0..sub_block_size)
+                            .into_par_iter()
+                            .for_each(|pos_in_sub_block| {
+                                let txn_idx = txn_idx_matrix[round_id][shard_id][pos_in_sub_block];
+                                *finalized_indexs[txn_idx].write().unwrap() =
+                                    start_index_matrix[round_id][shard_id] + pos_in_sub_block;
+                            });
+                    });
+            });
+        });
+        let duration = timer.stop_and_record();
+        info!("new_tid_table={duration}");
+        (txn_idx_matrix, start_index_matrix, finalized_indexs)
+    }
+}
+
+impl BlockPartitioner for PartitionerV2 {
+    fn partition(
+        &self,
+        txns: Vec<AnalyzedTransaction>,
+        num_executor_shards: usize,
+    ) -> Vec<SubBlocksForShard<AnalyzedTransaction>> {
+        let timer = MISC_TIMERS_SECONDS
+            .with_label_values(&["preprocess"])
+            .start_timer();
+        let num_txns = txns.len();
+        let num_senders = AtomicUsize::new(0);
+        let num_keys = AtomicUsize::new(0);
+        let mut senders: Vec<RwLock<Option<SenderIdx>>> = Vec::with_capacity(num_txns);
```

Comment:
Can you move the building index part to a separate function so that it makes this function a bit shorter and easier to follow?